src/shared/components/home/emojis-form.tsx

```diff
@@ -279,37 +298,80 @@ export class EmojiForm extends Component<EmojiFormProps, EmojiFormState> {
-    return noEmptyFields && noDuplicateShortCodes && !cv.loading && cv.changed;
+  canSave(cv: EditableEmoji) {
+    const requiredFields =
+      cv.emoji.custom_emoji.image_url.length > 0 &&
+      cv.emoji.custom_emoji.shortcode.length > 0;
+    return requiredFields && !cv.loading;
   }
 
-  getEditTooltip(cv: CustomEmojiViewForm) {
-    if (this.canEdit(cv)) return I18NextService.i18n.t("save");
+  getEditTooltip(cv: EditableEmoji) {
+    if (this.canSave(cv)) return I18NextService.i18n.t("save");
     else return I18NextService.i18n.t("custom_emoji_save_validation");
   }
 
-  handlePageChange(page: number) {
-    this.setState({ page: page });
+  async handlePageChange(page: number) {
+    this.setState({ loading: true });
+    let allEmojis: CustomEmojiView[] = this.state.allEmojis;
+    let emojiMartCustom: EmojiMartCategory[] = this.state.emojiMartCustom;
+    let emojiMartKey: number = this.state.emojiMartKey;
+    if (this.needsRefetch) {
+      const emojiRes = await HttpService.client.listCustomEmojis({
+        ignore_page_limits: true,
+      });
+      if (emojiRes.state === "success") {
+        this.needsRefetch = false;
+        allEmojis = emojiRes.data.custom_emojis;
+        allEmojis.sort((a, b) => {
+          if (a.custom_emoji.category < b.custom_emoji.category) {
+            return -1;
+          }
+          if (a.custom_emoji.category > b.custom_emoji.category) {
+            return 1;
+          }
+          if (a.custom_emoji.shortcode < b.custom_emoji.shortcode) {
+            return -1;
+          }
+          if (a.custom_emoji.shortcode > b.custom_emoji.shortcode) {
+            return 1;
+          }
+          return 0;
+        });
+      }
+      emojiMartCustom = emojiMartCategories(allEmojis);
+      emojiMartKey++;
+    }
+    if (allEmojis) {
+      const startIndex = (page - 1) * this.itemsPerPage;
+      const emojis = allEmojis
+        .slice(startIndex, startIndex + this.itemsPerPage)
+        .map(x => ({ emoji: x }));
+      this.setState({
+        loading: false,
+        allEmojis,
+        emojiMartCustom,
+        emojiMartKey,
+        emojis,
+        page,
+      });
+    } else {
+      this.setState({ loading: false, page });
+    }
   }
 
-  handleEmojiClick(e: any) {
-    const view = customEmojisLookup.get(e.id);
-    if (view) {
-      const page = this.state.customEmojis.find(
-        x => x.id === view.custom_emoji.id,
-      )?.page;
-      if (page) {
-        this.setState({ page: page });
-        this.scrollRef[view.custom_emoji.shortcode].scrollIntoView();
+  async handleEmojiClick(e: any) {
+    const emojiIndex = this.state.allEmojis.findIndex(
+      x => x.custom_emoji.shortcode === e.id,
+    );
+    if (emojiIndex >= 0) {
+      const { shortcode } = this.state.allEmojis[emojiIndex].custom_emoji;
+      const page = Math.floor(emojiIndex / this.itemsPerPage) + 1;
+      await this.handlePageChange(page);
+      await new Promise(r => setTimeout(r));
```

Comment:
What is this line doing?

Reply:
The timeout gives inferno enough time to add the new elements to the DOM.